File: 016. Binary Search/BinarySearchRecursive.java
import java.util.*;

class BinarySearchRecursive {
    public static int binarySearch(int[] array, int target) {
        // Write your code here.
        return binarySearch(array, target, 0, array.length - 1);
    }

    public static int binarySearch(int[] array, int target, int left, int right) {
        if (left > right) {
            return -1;
        }
        int pivot = (left + right) / 2;
        if (array[pivot] == target) {
            return pivot;
        } else if (target > array[pivot]) {
            return binarySearch(array, target, pivot + 1, right);
        } else {
            return binarySearch(array, target, left, pivot - 1);
        }
    }
}

File: 16. Binary Search/BinarySearchRecursive.kt
package com.algoexpert.program

fun binarySearch(array: List<Int>, target: Int): Int {
    // Write your code here.
    return binarySearch(array, target, 0, array.size - 1)
}

fun binarySearch(array: List<Int>, target: Int, left: Int, right: Int): Int {
    if (left > right) {
        return -1
    }
    var pivot: Int = (left + right) / 2
    if (array[pivot] == target) {
        return pivot
    } else if (array[pivot] < target) {
        return binarySearch(array, target, pivot + 1, right)
    } else {
        return binarySearch(array, target, left, pivot - 1)
    }
}
